File: test/operator/test_su2_hamiltonian.py
import numpy as np
import h5py
import sys
sys.path.append("../tensor/")
sys.path.append("../util/")
from hamiltonian import (
    construct_heisenberg_xxx_1d_hamiltonian,
    construct_fermi_hubbard_1d_hamiltonian)


def heisenberg_1d_su2_mpo_data():

    # number of lattice sites
    nsites = 5

    # Hamiltonian parameters
    J = 11./7

    # reference Hamiltonian
    heisenberg_1d_mat = construct_heisenberg_xxx_1d_hamiltonian(nsites, J).todense()

    with h5py.File("data/test_heisenberg_1d_su2_mpo.hdf5", "w") as file:
        file["heisenberg_1d_mat"] = heisenberg_1d_mat


def fermi_hubbard_1d_su2_mpo_data():

    # number of lattice sites
    nsites = 4

    # Hamiltonian parameters
    t  = 17./13
    u  = 7./3
    mu = -5./2

    # reference Hamiltonian
    fermi_hubbard_1d_mat = construct_fermi_hubbard_1d_hamiltonian(nsites, t, u, mu).toarray()
    # permutation of the local site basis states to SU(2) ordering (|00>, |11>, |10>, |01>)
    site_perm = np.array([
        [1., 0., 0., 0.],
        [0., 0., 0., 1.],
        [0., 0., 1., 0.],
        [0., 1., 0., 0.]])
    fermi_hubbard_1d_mat = fermi_hubbard_1d_mat.reshape((2*nsites) * (4,))
    for i in range(2*nsites):
        # multiply the i-th axis with 'site_perm'
        mat_idx = list(range(2*nsites))
        mat_idx[i] = 2*nsites
        fermi_hubbard_1d_mat = np.einsum(site_perm, (i, 2*nsites), fermi_hubbard_1d_mat, mat_idx, range(2*nsites))
    fermi_hubbard_1d_mat = fermi_hubbard_1d_mat.reshape(2 * (4**nsites,))

    with h5py.File("data/test_fermi_hubbard_1d_su2_mpo.hdf5", "w") as file:
        file.attrs["nsites"] = nsites
        file.attrs["t"]      = t
        file.attrs["u"]      = u
        file.attrs["mu"]     = mu
        file["fermi_hubbard_1d_mat"] = fermi_hubbard_1d_mat


def main():
    heisenberg_1d_su2_mpo_data()
    fermi_hubbard_1d_su2_mpo_data()


if __name__ == "__main__":
    main()

File: test/operator/test_su2_hamiltonian.c
#include "su2_hamiltonian.h"
#include "hdf5_util.h"


char* test_heisenberg_1d_su2_mpo()
{
	hid_t file = H5Fopen("../test/operator/data/test_heisenberg_1d_su2_mpo.hdf5", H5F_ACC_RDONLY, H5P_DEFAULT);
	if (file < 0) {
		return "'H5Fopen' in test_heisenberg_1d_su2_mpo failed";
	}

	// number of lattice sites
	const int nsites = 5;

	// Hamiltonian parameters
	const double J = 11./7;

	struct su2_mpo heisenberg_1d_mpo;
	construct_heisenberg_1d_su2_mpo(nsites, J, &heisenberg_1d_mpo);
	if (!su2_mpo_is_consistent(&heisenberg_1d_mpo)) {
		return "internal consistency check for SU(2) symmetric Heisenberg Hamiltonian MPO failed";
	}

	// convert to a matrix
	struct dense_tensor heisenberg_1d_mat;
	{
		struct su2_tensor heisenberg_1d_tensor;
		su2_mpo_to_tensor(&heisenberg_1d_mpo, &heisenberg_1d_tensor);
		if (heisenberg_1d_tensor.ndim_logical != 2 * nsites + 2) {
			return "expecting the SU(2) tensor representation of an SU(2) MPO to have logical degree '2 * nsites + 2'";
		}
		su2_to_dense_tensor(&heisenberg_1d_tensor, &heisenberg_1d_mat);
		delete_su2_tensor(&heisenberg_1d_tensor);
		// reshape into a matrix
		const ct_long hspace_dim = ipow(2, nsites);
		const ct_long dim[2] = { hspace_dim, hspace_dim };
		reshape_dense_tensor(2, dim, &heisenberg_1d_mat);
	}

	// reference matrix for checking
	struct dense_tensor heisenberg_1d_mat_ref;
	{
		hsize_t dim_hsize[2];
		if (get_hdf5_dataset_dims(file, "heisenberg_1d_mat", dim_hsize) < 0) {
			return "obtaining dimensions of reference Hamiltonian failed";
		}
		const ct_long dim[2] = { dim_hsize[0], dim_hsize[1] };
		allocate_dense_tensor(CT_DOUBLE_REAL, 2, dim, &heisenberg_1d_mat_ref);
		// read values from disk
		if (read_hdf5_dataset(file, "heisenberg_1d_mat", H5T_NATIVE_DOUBLE, heisenberg_1d_mat_ref.data) < 0) {
			return "reading matrix entries from disk failed";
		}
	}

	// compare
	if (!dense_tensor_allclose(&heisenberg_1d_mat, &heisenberg_1d_mat_ref, 1e-13)) {
		return "matrix representation of SU(2) symmetric Heisenberg Hamiltonian based on MPO form does not match reference";
	}

	// clean up
	delete_dense_tensor(&heisenberg_1d_mat_ref);
	delete_dense_tensor(&heisenberg_1d_mat);
	delete_su2_mpo(&heisenberg_1d_mpo);

	H5Fclose(file);

	return 0;
}
